rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs

```diff
@@ -11,26 +11,135 @@
 //! [`ConfigBag`](aws_smithy_types::config_bag::ConfigBag) instead of in
 //! [`RuntimeComponents`](RuntimeComponents).
 
+use crate::box_error::BoxError;
 use crate::client::auth::{
     AuthScheme, AuthSchemeId, ResolveAuthSchemeOptions, SharedAuthScheme,
     SharedAuthSchemeOptionResolver,
 };
 use crate::client::endpoint::{ResolveEndpoint, SharedEndpointResolver};
 use crate::client::http::{HttpClient, SharedHttpClient};
 use crate::client::identity::{
-    ConfiguredIdentityResolver, ResolveIdentity, SharedIdentityResolver,
+    ConfiguredIdentityResolver, ResolveCachedIdentity, ResolveIdentity, SharedIdentityCache,
+    SharedIdentityResolver,
 };
 use crate::client::interceptors::{Intercept, SharedInterceptor};
 use crate::client::retries::classifiers::{ClassifyRetry, SharedRetryClassifier};
 use crate::client::retries::{RetryStrategy, SharedRetryStrategy};
+use crate::impl_shared_conversions;
 use crate::shared::IntoShared;
 use aws_smithy_async::rt::sleep::{AsyncSleep, SharedAsyncSleep};
 use aws_smithy_async::time::{SharedTimeSource, TimeSource};
+use aws_smithy_types::config_bag::ConfigBag;
 use std::fmt;
+use std::sync::Arc;
 
 pub(crate) static EMPTY_RUNTIME_COMPONENTS_BUILDER: RuntimeComponentsBuilder =
     RuntimeComponentsBuilder::new("empty");
 
+/// Validates client configuration.
+///
+/// This trait can be used to validate that certain required components or config values
+/// are available, and provide an error with helpful instructions if they are not.
+pub trait ValidateConfig: fmt::Debug + Send + Sync {
+    /// Validate the base client configuration.
+    ///
+    /// This gets called upon client construction. The full config may not be available at
+    /// this time (hence why it has [`RuntimeComponentsBuilder`] as an argument rather
+    /// than [`RuntimeComponents`]). Any error returned here will become a panic
+    /// in the client constructor.
+    fn validate_base_client_config(
+        &self,
+        runtime_components: &RuntimeComponentsBuilder,
+        cfg: &ConfigBag,
+    ) -> Result<(), BoxError> {
+        let _ = (runtime_components, cfg);
+        Ok(())
+    }
+
+    /// Validate the final client configuration.
+    ///
+    /// This gets called immediately after the [`Intercept::read_before_execution`] trait hook
+    /// when the final configuration has been resolved. Any error returned here will
+    /// cause the operation to return that error.
+    fn validate_final_config(
+        &self,
+        runtime_components: &RuntimeComponents,
+        cfg: &ConfigBag,
+    ) -> Result<(), BoxError> {
+        let _ = (runtime_components, cfg);
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+enum ValidatorInner {
+    BaseConfigStaticFn(fn(&RuntimeComponentsBuilder, &ConfigBag) -> Result<(), BoxError>),
```

Comment:
not sure it's worth having the two variants—I think you could turn one into the other during construction?

doesn't really matter anyway though—it's private.

Reply:
I was just avoiding unnecessary allocation.